src/kaleidoscope/hid.cpp

```diff
@@ -51,13 +51,86 @@
 namespace kaleidoscope {
 namespace hid {
 
+// Anonymous namespace for internal variables and helper functions
+namespace {
+
+// The functions in this namespace are primarily to solve the problem of rollover from a
+// key with a modifier flag (e.g. `LSHIFT(Key_T)`) to one without (e.g. `Key_H`), which
+// used to result in the mod flag being applied to keys other than the one with the
+// flag. By using `mod_flags_allowed`, we can prohibit any modifier flags that aren't in
+// the key most recently pressed, fixing the rollover problem, and getting the intended
+// `The` instead of `THe`. This was particularly noticeable to people who were using a
+// separate layer for `shift`, and a OneShot layer key to activate it, causing what looked
+// like a OneShot bug.
+
+// A bitfield of the current modifier flags from keys processed so far in the current
+// keyswitch scan:
+static byte current_mod_flags{0};
+
+// The current set of modifier flags that will be allowed to be added to the upcoming
+// keyboard HID report. This should be set by the latest key pressed:
+static byte mod_flags_allowed{0};
+
+// The keycode of the last key to toggle on. This get
+static byte toggled_on_keycode{0};
+
+// This returns true if the key is a keyboard key, and its keycode is a modifier.
+bool isPureModifier(Key mappedKey) {
```

Comment:
Nice PR. Very well commented!

As `isPureModifier`, `addModFlags` and `pressModFlags` are only called in one place, it might be beneficial in terms of code size to mark them `inline`. But maybe the compiler is already smart enough.

Reply:
Good point; thanks. I actually thought of this momentarily last night, when writing up comments for one of those functions, but then promptly forgot the idea.

Reply:
I just checked, and the binary is the same – the first two get inlined, and `pressModFlags()` doesn't, whether or not I declare them `inline` (`pressModFlags()` actually gets called in two places, and it's clearly the biggest of the three, so that makes sense to me). @noseglasses, would you still recommend explicitly declaring them `inline`, regardless? Maybe just the first two?

Reply:
That's kind of what I suspected, although I was not sure.

The compiler does the right thing by inlining the first two functions but not the latter (as called in two places). Because you put it in the anonymous namespace the functions are unavailable to other compilation units. That's why no additional actual function binary code is generated even though the functions are not inlined.

In that case it is not worth the hassle to inline.

Reply:
instead of isPureModifier, maybe breaking it into 

isModifier() which does the modifier key keycode check 

and isSynthetic() which does the other check?

Reply:
I'd been thinking of a negated check for `isSynthetic()`
```
if (isModifier(mappedKey) && ! isSynthetic(mappedKey)) {

}
```

I suppose the positive version would be `isKeyboardKey()`, maybe? that may be a little more understandable